contract/RadicleCloudBase.sol

```diff
@@ -0,0 +1,191 @@
+    function topUp(address org) payable public {
+        if (msg.value >= pricePerMonth) {
+            if (expiry[org] > block.number) {
+                // active user
+                expiry[org] = expiry[org] + blocksPerMonth;
+            } else {
+                // new user
+                expiry[org] = uint64(block.number + blocksPerMonth);
+            }
+            emit NewTopUp(org, expiry[org]);
+
+            // create a new head, increase shares and link previous head to current one
+            SubStatus memory newHeadStatus = subStatus[lastCreatedSubStatus];
+            newHeadStatus.shares += blocksPerMonth;
+            //newHeadStatus.previous = lastCreatedSubStatus;
+
+            uint64 currBlock = uint64(block.number);
+            subStatus[lastCreatedSubStatus].next = type(uint64).max; // currBlock;
+            lastCreatedSubStatus = currBlock;
+            subStatus[currBlock] = newHeadStatus;
+        }
+    }
+
+    /// @dev Returns smaller number between 'a' and 'b'.
+    function min(uint64 a, uint64 b) internal returns (uint64) {
+        return a < b ? a : b;
+    }
+
+    /// @dev Returns minimum between last buy, expiry, cancel or offset event
+    function _min_of_four() internal returns(uint64) {
+        return min(min(buys[LAST[BUY]], expires[LAST[EXPIRE]]), min(cancels[LAST[CANCEL]], offsets[LAST[OFFSET]]));
+    }
+
+    /// @dev Calculates the amount operator has earned.
+    /// NOTE: BUY/EXPIRE is ordered by nature of chain, CANCEL/OFFSET needs sorting on insert
+    function _calcRealizedRevenue() internal returns (int128)  {
+        uint128 revenue = 0;
+
+        uint64 last = LAST[PROCESSED];
+        uint64 currBlock = uint64(block.number);
+        uint64 rate = RATE;
+
+        while (last < currBlock) {
+            uint64 next_last = _min_of_four();
+
+            if (next_last == MAXBLOCK) {
+                LAST[PROCESSED] = currBlock;
+                break;
+            }
+
+            if (last == next_last) {
+                while (last == next_last) {
+                    if (last == buys[LAST[BUY]]) {
+                        rate += 1;
+                        LAST[BUY] += 1;
+                    }
+                    if (last == expires[LAST[EXPIRE]]) {
+                        rate -= 1;
+                        LAST[EXPIRE] += 1;
+                    }
+                    if (last == cancels[LAST[CANCEL]]) {
+                        rate -= 1;
+                        LAST[CANCEL] += 1;
+                    }
+                    if (last == offsets[LAST[OFFSET]]) {
+                        rate += 1;
+                        LAST[OFFSET] += 1;
+                    }
+                    next_last = _min_of_four();
+                }
+                continue;
+            }
+            revenue += rate * (min(currBlock, next_last) - last);
+            last = next_last;
+        }
+
+        RATE = rate;
+        return int128(revenue);
```

Comment:
How can the revenue be negative?

Reply:
Drips uses int128 https://github.com/radicle-dev/radicle-drips-hub/blob/master/src/DripsHub.sol#L785 , should I change it to uint128?